components/renku_data_services/session/core.py

```diff
@@ -238,23 +239,26 @@ def validate_session_launcher_patch(
                         message="There are errors in the following fields, environment.build_parameters: Must be null"
                     )
                 elif (
+                    # TODO: Add a test for new == None/"build" and current == "build"
                     new == "build" or (new is None and current == "build")
                 ) and patch.environment.build_parameters is None:
                     raise errors.ValidationError(
                         message="There are errors in the following fields, environment.build_parameters: Must be set"
                     )
+
                 if current == "image" and new == "build":
-                    if not patch.environment.build_parameters:
-                        raise errors.ValidationError(
-                            message=(
-                                "There are errors in the following fields, environment.build_parameters: Must be set"
-                            )
-                        )
+                    assert patch.environment.build_parameters is not None  # Make mypy happy in the following statement
+                    # NOTE: The environment type is changed to be built, so, all required fields should be passed (as in
+                    # a POST request). No need to get values from the current env, since they will be set by the build.
```

Comment:
I'm not sure assert is the right thing to do here. if this were to happen, it'd trigger an assertion error which probably isn't handled correctly by our error handling.

Is using `if patch.environment.build_parameters is None:` with the raise that we had before not enough to keep mypy happy (like changing the `not ...` to `is None`)?

Reply:
We are already checking that `patch.environment.build_parameters` is not `None` in the previous if block and this check was redundant. However, `mypy` isn't smart enough to deduct that `patch.environment.build_parameters` is not null. I'm not sure what's the general policy to handle such cases, but we've used assertion before. I can re-introduce the if/raise block if you think that's a better way of handling it (maybe it's more readable).